aesd-char-driver/main.c

```diff
@@ -4,73 +4,130 @@
  *
  * Based on the implementation of the "scull" device driver, found in
  * Linux Device Drivers example code.
- *
- * @author Dan Walkes
- * @date 2019-10-22
- * @copyright Copyright (c) 2019
- *
  */
 
 #include <linux/module.h>
 #include <linux/init.h>
 #include <linux/printk.h>
 #include <linux/types.h>
 #include <linux/cdev.h>
-#include <linux/fs.h> // file_operations
+#include <linux/fs.h>
+#include <linux/uaccess.h>
+#include <linux/slab.h>
 #include "aesdchar.h"
-int aesd_major =   0; // use dynamic major
-int aesd_minor =   0;
 
-MODULE_AUTHOR("Your Name Here"); /** TODO: fill in your name **/
+int aesd_major = 0;  // use dynamic major
+int aesd_minor = 0;
+
+MODULE_AUTHOR("Your Name");
 MODULE_LICENSE("Dual BSD/GPL");
 
 struct aesd_dev aesd_device;
 
 int aesd_open(struct inode *inode, struct file *filp)
 {
     PDEBUG("open");
-    /**
-     * TODO: handle open
-     */
+    filp->private_data = &aesd_device;
     return 0;
 }
 
 int aesd_release(struct inode *inode, struct file *filp)
 {
     PDEBUG("release");
-    /**
-     * TODO: handle release
-     */
     return 0;
 }
 
-ssize_t aesd_read(struct file *filp, char __user *buf, size_t count,
-                loff_t *f_pos)
+ssize_t aesd_read(struct file *filp, char __user *buf, size_t count, loff_t *f_pos)
 {
+    struct aesd_dev *dev = filp->private_data;
+    struct aesd_buffer_entry *entry;
+    size_t entry_offset;
+    size_t bytes_to_copy;
     ssize_t retval = 0;
-    PDEBUG("read %zu bytes with offset %lld",count,*f_pos);
-    /**
-     * TODO: handle read
-     */
+
+    PDEBUG("read %zu bytes with offset %lld", count, *f_pos);
+
+    mutex_lock(&dev->lock);
+
+    entry = aesd_circular_buffer_find_entry_offset_for_fpos(&dev->buffer, *f_pos, &entry_offset);
+    if (!entry) {
+        mutex_unlock(&dev->lock);
+        return 0; /* EOF */
+    }
+
+    bytes_to_copy = min(entry->size - entry_offset, count);
+
+    if (copy_to_user(buf, entry->buffptr + entry_offset, bytes_to_copy)) {
+        mutex_unlock(&dev->lock);
+        return -EFAULT;
+    }
+
+    *f_pos += bytes_to_copy;
+    retval = bytes_to_copy;
+
+    mutex_unlock(&dev->lock);
     return retval;
 }
 
-ssize_t aesd_write(struct file *filp, const char __user *buf, size_t count,
-                loff_t *f_pos)
+ssize_t aesd_write(struct file *filp, const char __user *buf, size_t count, loff_t *f_pos)
 {
-    ssize_t retval = -ENOMEM;
-    PDEBUG("write %zu bytes with offset %lld",count,*f_pos);
-    /**
-     * TODO: handle write
-     */
+    struct aesd_dev *dev = filp->private_data;
+    struct aesd_buffer_entry new_entry;
+    char *new_buf;
+    ssize_t retval = count;
+    bool newline_found = false;
+    size_t i;
+
+    PDEBUG("write %zu bytes with offset %lld", count, *f_pos);
+
+    new_buf = kmalloc(count + dev->partial_size, GFP_KERNEL);
```

Comment:
@parthnpatel20 is it safe to access dev->partial_size here without holding the mutex?